src/sentry/notifications/notificationcontroller.py

```diff
@@ -0,0 +1,431 @@
+            scope_identifier__in=user_ids,
+        )
+
+        return project_settings | org_settings | team_or_user_settings
+
+    def get_scope_for_layering(
+        self, recipient: RpcActor | Team | RpcUser
+    ) -> Tuple[NotificationScopeEnum, Iterable[int | RpcActor | Team | RpcUser]]:
+        if self.project_ids:
+            return (NotificationScopeEnum.PROJECT, self.project_ids)
+        if self.organization_id is not None:
+            return (NotificationScopeEnum.ORGANIZATION, [self.organization_id])
+        if recipient_is_user(recipient):
+            return (NotificationScopeEnum.USER, [recipient])
+        if recipient_is_team(recipient):
+            return (NotificationScopeEnum.TEAM, [recipient])
+
+        raise ValueError(f"Recipient is of unsupported type: {recipient}")
+
+    def get_layered_setting_options(
+        self,
+        setting_type: NotificationSettingEnum | None = None,
+        additional_filters: Q | None = None,
+    ) -> MutableMapping[
+        Recipient,
+        MutableMapping[
+            Scope, MutableMapping[NotificationSettingEnum, NotificationSettingsOptionEnum]
+        ],
+    ]:
+        """
+        Returns a mapping of the most specific notification setting options for the given recipients and scopes.
+        Note that this includes default settings for any notification types that are not set.
+
+        Args:
+            setting_type: If specified, only return settings of this type.
+            additional_filters: Additional filters to apply to the query.
+        """
+        if not additional_filters:
+            additional_filters = Q()
+
+        if setting_type:
+            additional_filters &= Q(type=setting_type.value)
+
+        settings = self.get_all_setting_options().filter(additional_filters)
+        project_settings = settings.filter(scope_type=NotificationScopeEnum.PROJECT.value)
+        org_settings = settings.filter(scope_type=NotificationScopeEnum.ORGANIZATION.value)
+        user_settings = settings.filter(scope_type=NotificationScopeEnum.USER.value)
+        scoped_settings = [project_settings, org_settings, user_settings]
+
+        defaults = get_type_defaults()
+        layered_setting_options: MutableMapping[
+            Recipient,
+            MutableMapping[
+                Scope, MutableMapping[NotificationSettingEnum, NotificationSettingsOptionEnum]
+            ],
+        ] = defaultdict(lambda: defaultdict(dict))
+        for recipient in self.recipients:
+            (scope_type, scope_items) = self.get_scope_for_layering(recipient)
+            user_or_team_filter = Q(user_id=recipient.id) | Q(team_id=recipient.id)
+            for item in scope_items:
+                scope: Scope = (scope_type, item)
+                for type in NotificationSettingEnum:
+                    most_specific_setting = None
+                    for setting in scoped_settings:
+                        s = setting.filter(user_or_team_filter, type=type.value)
```

Comment:
@snigdhas isn't this still an n+1 query? Might be better to load all the settings in memory in a list and then filter in-memory rather than as a layered Django query.